lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -502,6 +502,122 @@ static StringRef getOperandName(OpOperand &oper, const hw::SymbolCache &syms,
+      auto refSymbol = globalRef.getGlblSym();
+      auto globalRefOp = rootModule.lookupSymbol<hw::GlobalRefOp>(refSymbol);
+      auto oldPath = globalRefOp.namepath().getAsRange<hw::InnerRefAttr>();
+
+      // Construct a new path starting from the old path.
+      SmallVector<Attribute> newPath(oldPath);
+
+      // If there aren't two elements in the path, this GlobalRefOp has already
+      // been updated.
+      if (newPath.size() < 2)
+        return;
+
+      // Splice the last two elements in the path, combining their names with a
+      // dot, and taking the module from the second to last element.
+      auto lastElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
+      auto nextElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
+      auto newModule = nextElement.getModule();
+
+      SmallString<16> newName;
+      newName += nextElement.getName().getValue();
+      newName += '.';
+      newName += lastElement.getName().getValue();
+      auto newNameAttr = StringAttr::get(op->getContext(), newName);
+
+      auto newLeaf = hw::InnerRefAttr::get(newModule, newNameAttr);
+      newPath.push_back(newLeaf);
+
+      // Update the path on the GlobalRefOp.
+      auto newPathAttr = ArrayAttr::get(op->getContext(), newPath);
+      globalRefOp.namepathAttr(newPathAttr);
+    }
+  }
+}
+
+/// Helper to update GlobalRefops after referenced ops are pushed down.
+static void pushDownGlobalRefs(Operation *op) {
+  if (auto globalRefs = op->getAttrOfType<ArrayAttr>("circt.globalRef")) {
+    for (auto globalRef : globalRefs.getAsRange<hw::GlobalRefAttr>()) {
+      // Resolve the GlobalRefOp and get its path.
+      auto rootModule = op->getParentOfType<ModuleOp>();
+      auto refSymbol = globalRef.getGlblSym();
+      auto globalRefOp = rootModule.lookupSymbol<hw::GlobalRefOp>(refSymbol);
+      auto oldPath = globalRefOp.namepath().getAsRange<hw::InnerRefAttr>();
+
+      // Get the module containing the partition and the partition's name.
+      auto partAttr = op->getAttrOfType<SymbolRefAttr>("targetDesignPartition");
+      auto partMod = partAttr.getRootReference();
+      auto partName = partAttr.getLeafReference();
+
+      // Resolve the DesignPartitionOp and get its verilog name.
+      auto mod = op->getParentOfType<MSFTModuleOp>();
+      StringAttr partModName;
+      for (auto part : mod.getOps<DesignPartitionOp>()) {
+        if (part.sym_nameAttr() == partName) {
+          partModName = part.verilogNameAttr();
+          break;
+        }
+      }
+      assert(partModName);
+
+      // Construct a new path starting from the old path.
+      SmallVector<Attribute> newPath(oldPath);
+      auto lastElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
+
+      // If the last element in the path is not pointing at the module
+      // containing the partition, this GlobalRefOp has already been updated.
+      if (lastElement.getModule() != partMod)
```

Comment:
Assert?

Reply:
This is the same as in bubbleUp: this condition is true for some intermediate nodes, but when it is false, we want to continue executing the body of this block.